source/h2_frames.c

```diff
@@ -332,714 +256,829 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
+static struct aws_h2_frame_prebuilt *s_h2_frame_new_prebuilt(
+    struct aws_allocator *allocator,
+    enum aws_h2_frame_type type,
+    uint32_t stream_id,
+    size_t payload_len,
+    uint8_t flags) {
+
+    AWS_PRECONDITION(payload_len <= s_prebuilt_payload_max());
+
+    const size_t encoded_frame_len = s_frame_prefix_length + payload_len;
+
+    /* Use single allocation for frame and buffer storage */
+    struct aws_h2_frame_prebuilt *frame;
+    void *storage;
+    if (!aws_mem_acquire_many(
+            allocator, 2, &frame, sizeof(struct aws_h2_frame_prebuilt), &storage, encoded_frame_len)) {
+        return NULL;
+    }
 
-int aws_h2_frame_goaway_encode(
-    struct aws_h2_frame_goaway *frame,
-    struct aws_h2_frame_encoder *encoder,
-    struct aws_byte_buf *output) {
-    AWS_PRECONDITION(frame);
-    AWS_PRECONDITION(encoder);
-    AWS_PRECONDITION(output);
+    AWS_ZERO_STRUCT(*frame);
+    s_init_frame_base(&frame->base, allocator, type, &s_frame_prebuilt_vtable, stream_id);
+    frame->encoded_buf = aws_byte_buf_from_empty_array(storage, encoded_frame_len);
+    frame->send_progress = aws_byte_cursor_from_array(storage, encoded_frame_len);
 
-    (void)encoder;
+    /* Write frame prefix */
+    s_frame_prefix_encode(type, stream_id, payload_len, flags, &frame->encoded_buf);
 
-    if (frame->base.stream_id != 0) {
-        return aws_raise_error(AWS_ERROR_HTTP_PROTOCOL_ERROR);
-    }
+    return frame;
+}
 
-    const size_t output_init_len = output->len;
+static void s_frame_prebuilt_destroy(struct aws_h2_frame *frame_base) {
+    aws_mem_release(frame_base->alloc, frame_base);
+}
 
-    /* Write the header data */
-    size_t length = 8 + frame->debug_data.len;
-    if (s_frame_prefix_encode(&frame->base, length, 0, output)) {
-        goto write_error;
+static int s_frame_prebuilt_encode(
+    struct aws_h2_frame *frame_base,
+    struct aws_h2_frame_encoder *encoder,
+    struct aws_byte_buf *output,
+    bool *complete) {
+
+    struct aws_h2_frame_prebuilt *frame = AWS_CONTAINER_OF(frame_base, struct aws_h2_frame_prebuilt, base);
+
+    if (frame->send_progress.len == frame->encoded_buf.len) {
+        ENCODER_LOGF(
+            TRACE,
+            encoder,
+            "Encoding frame type=%s stream_id=%" PRIu32 " - begin",
+            aws_h2_frame_type_to_str(frame->base.type),
+            frame->base.stream_id);
     }
 
-    /* Write the payload */
-    if (!aws_byte_buf_write_be32(output, frame->last_stream_id & s_31_bit_mask)) {
-        goto write_error;
-    }
-    if (!aws_byte_buf_write_be32(output, frame->error_code)) {
-        goto write_error;
-    }
-    if (!aws_byte_buf_write_from_whole_cursor(output, frame->debug_data)) {
-        goto write_error;
-    }
+    bool all_wrote = true;
 
-    return AWS_OP_SUCCESS;
+    /* Write as much of the pre-encoded frame as will fit */
+    size_t chunk_len = aws_min_size(frame->send_progress.len, output->capacity - output->len);
```

Comment:
why isn't this ALWAYS zero? I think Justin and I had the same question... maybe some comments on this somewhere for future posterity?

Reply:
I changed some variable names and added comments.
Hope this is less confusing now